### DISCUSSION TITLE
New Behave version ?

### DISCUSSION THREAD

**Discussion by nad56** [behave/behave#1030] [General]
Do you plan soon to release a new version?
Indeed, we need to package behave for fedora and for that, we would need to have behave without nose dependency (as it is already done in the main trunk).

Thanks!

**Comment by bittner** [comment]
In relation to https://github.com/behave/behave/issues/855#issuecomment-994772325 I'd suggest to package up a 1.3.0 version as the last version to support Python 2.7, and then drop Python 2.7 support for a new version 2.0.0 to come.

This may then mark the beginning of a semantic versioning scheme adherence, with potentially more frequent, largely automated releases.

@jenisys Any comment on this?

**Comment by HenriBlacksmith** [comment]
Any plans to release 1.2.7 or 1.3.0 (or 2.0.0)?

Seems that `pyproject.toml` support was added, would be a nice feature to release.

**Comment by pauloxnet** [comment]
Honestly 5 years waiting for a new release is too much.
Are there any issues that really block the release of a new version?

**Comment by rkmullapudi** [comment]
Hi, yes, looking for new release. Please let us know if there is plan to release new version sometime soon?

**Comment by HenriBlacksmith** [comment]
I am wondering about the release strategy for behave?

The project is still maintained by @jenisys but I assume the master branch is not stable enough to release something?

Otherwise if atomicity of MR is guaranteed, I do not see any blocker to release?

**Comment by jenisys** [comment]
@HenriBlacksmith 
The `main` branch (not: `master`, that was the old branch name) is stable and tested with CI.
The Git repository contains a number of (pre-)release tags that you can use before an official release is publish on PyPI.
The installation recipe for using Github (tags or branch) is described in the documentation.

There are some issues pending that need to be resolved before the next official release is published on PyPI.

**Reply by bittner** [reply]
> There are some issues pending that need to be resolved before the next official release is published on PyPI.

Can we maybe jot down **a list of those issues** that need to be resolved to unblock a release? I believe, the interest of the community to get a release (actually, regular releases) out there is so big that it might be realistic to assume there will be contributors willing to go the extra mile (with PRs) just to make sure a release finally happens.

**Comment by mpinardi** [comment]
I am unsure of your timelines, but I have just completed a early version of a performance testing tool that sits on top of behave 1.2.7 (A java/js cucumber editions already exists). Could you possibility do 1.2.7-dev package release? Because as pipy does not allow direct references and as such I can't push my tool into the public repo. I was hopeful that as it took me 6 months to get this version done that you might have beaten me to the punch but that seems unlikely now.

**Reply by bittner** [reply]
I'm not 100% sure what you're referring to with "dev package release". If you intend "a release on PyPI" that will probably not happen as a dev package, according to my experience (not my choice).

OTOH, you can install the latest intended dev-release from this repository using any of the latest dev-tags ([see all tags](https://github.com/behave/behave/tags)). You can integrate this in your packaging just like references to packages on PyPI. Place a line like this one in your `pyproject.toml`, `setup.py`, `requirements.txt` or the like (see the [dependencies of behave-django](https://github.com/behave/behave-django/blob/main/requirements.txt#L2) for a real-world example):

```python
behave[toml]@git+http://github.com/behave/behave.git@v1.2.7.dev5
```

I know, this might not seem ideal for various reasons, but that's the best choice available today, I would guess. :woman_shrugging:

**Reply by mpinardi** [reply]
Yes I meant release dev packages to PyPi such as 1.2.7-dev5. 
But in regards to configuration in setup.py in the way mentioned.
PyPi  does not allow this either:
```
 Response from https://test.pypi.org/legacy/:
         400 Invalid value for requires_dist. Error: Can't have direct dependency: 'behave[toml] @
         git+http://github.com/behave/behave@v1.2.7.dev5'
```
So again it seems that you can not push a release to PyPi if you reference a dependency directly.

**Reply by bittner** [reply]
Can you show the (complete) source code of your packaging setup?

According to the error message, you're using `requires_dist`. Why not `install_requires`, like below?

```python
setup(
    name="foo-bar",
    version="1.2.3",
    install_requires=[
        "behave[toml]@git+http://github.com/behave/behave.git@v1.2.7.dev5",
    ],
    ...
)
```

This should work fine. Take a look at the [supported syntax](https://pip.pypa.io/en/stable/reference/requirement-specifiers/) in the Pip documentation, otherwise.

**Comment by mpinardi** [comment]
```
    ...
    python_requires='>=3.9',
    install_requires=['pyee','aiofiles','behave[toml] @ git+http://github.com/behave/behave.git@v1.2.7.dev5'],
    test_suite="test"`
```
result

`         <title>400 Invalid value for requires_dist. Error: Can't have direct dependency: 'behave[toml] @       
         git+http://github.com/behave/behave.git@v1.2.7.dev5'</title>
          </head>
          <body>`

**Reply by bittner** [reply]
Looks like you're right, because the behavior [appears to be by design](https://github.com/pypa/pip/issues/6301#issuecomment-533893126) ([2](https://github.com/pypi/warehouse/issues/7136#issuecomment-603529974)) on PyPI. Thanks for pointing this out!

@jenisys Could you push a development release to PyPI, for real? The problem also affects our project directly as I couldn't release a working state of [behave-django](https://github.com/behave/behave-django) on PyPI, which requires `behave@v1.2.7.dev4`. :bomb:

**Reply by bittner** [reply]
Related issue: https://github.com/behave/behave-django/issues/158

**Comment by Johnlon** [comment]
Re New release. 
I can't help but notice the lack of replies above to those asking what the release schedule is for behave.

Can we please have a reasponse here, or a link to a page, where the maintainers give a direct answer on the plans for behave.

I can only assume that, given the active commits, that the contributors are using dev builds or builds from their own forked copies?
Is that's what's going on?
If so then what is the incentive for the maintainers to make a formal release of something they consider stable?

For those of us who don't want to maintain a fork and our own build, should we consider behave as effectively dead?

That would be a pity - but we need something concrete after 6 years.

RSVP

**Reply by HenriBlacksmith** [reply]
I think @jenisys is the last active maintainer (correct me if I am wrong), I assume it is up to him to clarify the blockers which prevented him to release an actual tag all this time (and to see how contributors could help here)

**Comment by bittner** [comment]
@jenisys Can you give an update for everyone? As a starting point, we could start pushing all development releases to PyPI. Would that be feasible?

**Comment by Johnlon** [comment]
@jenisys is still merging changes but as far as I'm aware has still not responded to any queries (ever?) relating to releases.

even a markdown file message from @jenisys explaining that there won't be any further releases and that we should roll our own would be an improvement (wasn't there last time I looked)

Given the lack of responses its possibly past the point where this should be forked and new maintainers identified etc.
Maybe fork it into the cucumber code repo?
I recall  Matt Wynne asked same question in some other ticket or chat somewhere.

**Reply by Johnlon** [reply]
but unless @jenisys stops merging to trunk then we'll just end up with two code bases diverging. 

given the silence from @jenisys I don't feel hopeful and have abandoned all use of behave myself in favour of other tools

**Comment by jenisys** [comment]
`behave v1.2.7.dev6` was published on pypi (as pre-release):

* https://pypi.org/project/behave/1.2.7.dev6/

**Reply by Johnlon** [reply]
will there be a "release" version?

**Reply by jenisys** [reply]
Mmh, what should I say.
This is a release (even if it is just a pre-release as the `.dev6` suffix indicates).
This will eventually lead to the `v1.3.0` release.

SEE ALSO:
* Roadmap sketched out at the begin of [CHANGELOG](https://github.com/behave/behave/blob/main/CHANGES.rst).
